core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+    val info = get(blockId).get
+    require(info.writerTask == currentTaskAttemptId,
+      s"Task $currentTaskAttemptId tried to downgrade a write lock that it does not hold")
+    unlock(blockId)
+    lockForReading(blockId, blocking = false)
+    notifyAll()
+  }
+
+  /**
+   * Release a lock on the given block.
+   */
+  def unlock(blockId: BlockId): Unit = synchronized {
+    logTrace(s"Task $currentTaskAttemptId releasing lock for $blockId")
+    val info = get(blockId).getOrElse {
+      throw new IllegalStateException(s"Block $blockId not found")
+    }
+    if (info.writerTask != -1) {
+      info.writerTask = -1
+      writeLocksByTask.removeBinding(currentTaskAttemptId, blockId)
+    } else {
+      assert(info.readerCount > 0, s"Block $blockId is not locked for reading")
+      info.readerCount -= 1
+      val countsForTask = readLocksByTask.get(currentTaskAttemptId)
+      val newPinCountForTask: Int = countsForTask.remove(blockId, 1) - 1
+      assert(newPinCountForTask >= 0,
+        s"Task $currentTaskAttemptId release lock on block $blockId more times than it acquired it")
+    }
+    notifyAll()
+  }
+
+  /**
+   * Atomically create metadata for a non-existent block.
+   *
+   * @param blockId the block id.
+   * @param newBlockInfo the block info for the new block.
+   * @return true if the block did not already exist, false otherwise. If this returns false, then
+   *         no new locks are acquired. If this returns true, a write lock on the new block will
+   *         be held.
+   */
+  def lockNewBlockForWriting(
+      blockId: BlockId,
+      newBlockInfo: BlockInfo): Boolean = synchronized {
+    logTrace(s"Task $currentTaskAttemptId trying to put $blockId")
+    val actualInfo = infos.getOrElseUpdate(blockId, newBlockInfo)
+    if (actualInfo eq newBlockInfo) {
+      actualInfo.writerTask = currentTaskAttemptId
+      writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Release all lock held by the given task, clearing that task's pin bookkeeping
+   * structures and updating the global pin counts. This method should be called at the
+   * end of a task (either by a task completion handler or in `TaskRunner.run()`).
+   *
+   * @return the ids of blocks whose pins were released
+   */
+  def unlockAllLocksForTask(taskAttemptId: TaskAttemptId): Seq[BlockId] = {
+    val blocksWithReleasedLocks = mutable.ArrayBuffer[BlockId]()
+    synchronized {
+      writeLocksByTask.remove(taskAttemptId).foreach { locks =>
+        for (blockId <- locks) {
+          infos.get(blockId).foreach { info =>
+            assert(info.writerTask == taskAttemptId)
+            info.writerTask = -1
+          }
+          blocksWithReleasedLocks += blockId
+        }
+      }
+      notifyAll()
+    }
+    val readLocks = readLocksByTask.get(taskAttemptId)
+    readLocksByTask.invalidate(taskAttemptId)
```

Comment:
The assumption here is that this task's threads have finished executing, so it will not attempt to acquire any additional locks after this method has been called. If this assumption were not to hold, then we'd need to guard against new locks occurring after `unlockAllLocksForTask` returns. I think we might be able to handle this by requiring an explicit `registerTask()` to be called in order to initialize the map entries in `readLocksByTask` and `writeLocksByTask`; that would let us detect and reject invalid lock calls that occur after task completion. I'll take a shot at this tomorrow.